### ISSUE TITLE
1.0 Tracking issue

### ISSUE THREAD

**Issue by MabezDev** [rust-embedded/critical-section#14]
The token changes + emitting `bare_metal::CriticalSection` are really useful, is it worth pushing a 0.3 release now? Perhaps a 1.0 is better, to avoid fragmentation, if so do we have a known set of issues/features/blockers for 1.0?

**Comment by Dirbaio** [comment]
a few:

- There were talks of folding `bare-metal` into this crate. things have stalled a bit, so I think I'll just go ahead and do it myself: 
- Naming: the "restore token" name clashes with the "CriticalSection token" which is the actually user-facing one, so perhaps it needs a rename `RestoreState` maybe.

**Comment by Dirbaio** [comment]
I've fixed the above items, and released [`1.0.0-alpha.1`](https://crates.io/crates/critical-section/1.0.0-alpha.1).

**Comment by jannic** [comment]
I hope it's ok to use this ticket to report potential issues with the current alpha implementation. If you prefer, I can open a separate ticket, as well.

I think there is some issue with the upgrade path to 1.0. It seems possible to have dependencies on both 0.2 and 1.0 in the same project. (Which is great! Makes transition so much easier! If it works, reliably...)

Of course, you can call only one of `custom_impl!`/`set_impl!`, otherwise you get errors like `the name `_critical_section_release` is defined multiple times`. This is also fine for `restore-state-u8` as in that case the signatures of `_critical_section_acquire` and `_critical_section_release` match.

But for other types of restore state, the caller will use a different type than the callee. This _may_ also work by chance, but it's not guaranteed, and I guess this is UB.
In practice, I get completely fine assembly as the output, in this case an application using `0.2` with an implementation using critical-section 1.0.0-alpha.1, with `restore-state-u32`:
```
1000082e: 01 f0 77 fe   bl      0x10002520 <_critical_section_acquire> @ imm = #7406
10000832: 06 99         ldr     r1, [sp, #24]
10000834: 62 4a         ldr     r2, [pc, #392]          @ 0x100009c0 <$d.5+0x44>
10000836: 11 60         str     r1, [r2]
10000838: 01 f0 a4 fe   bl      0x10002584 <_critical_section_release> @ imm = #7496
```
The return value of `_critical_section_acquire` in `r0` is not even touched before calling `_critical_section_release`, and even if `r0` was used by the code in the critical section, the compiler would probably just push `r0` to the stack, so it would work for all types but `restore-state-u64`. Still, just because it happens to work, UB doesn't become correct.

With `restore-state-u64`, the calling code stays the same, but the state is returned / passed to release in `r0` and `r1`, which is overwritten by the code inside the critical section → the resulting binary is broken, without any warning from the compiler.

How can this be solved? One possibility would be changing the function names, so mixing both versions would cause an undefined symbol error. Or, if both of `custom_impl!`/`set_impl!` are called, it would compile, and work as expected, as long as the implementations are "compatible", ie. they properly lock each other out.

Another option would be making the symbol depend on the type of `RestoreState`, and keep the old name for `restore-state-u8`. 
That would make the versions compatible for this choice.

**Comment by Dirbaio** [comment]
ooh I hadn't thought about the 0.2 and 1.0 externs interacting. This is a problem, yes!

Sharing the externs (even in the u8 case only) could be problematic maybe. Since 0.2 will supply the default impls by default, if you have it anywhere in the tree you'll get dupllicate symbol errors.

**Comment by jannic** [comment]
Those duplicate impls can be fixed by still depending on the old version with the `custom-impl` feature enabled.
This seems to work fine: https://github.com/rp-rs/rp-hal/pull/403/files (Not heavily tested, though.)
Perhaps critical-section 1.0.0 should do that implicitly?

**Comment by Dirbaio** [comment]
Making CS1.0 depend on CS0.2 forever feels a bit hacky. Also, it'd only work with `restore-token-u8`.

CS0.2 isn't [that widely used](https://crates.io/crates/critical-section/reverse_dependencies) so IMO a clean break wouldn't be that bad.

**Comment by Dirbaio** [comment]
Changed name of externs, released 1.0.0-alpha.2

**Comment by jannic** [comment]
Isn't the state with 1.0.0-alpha.2 a little bit dangerous?
If your HAL (or whatever provides the implementation) switches to 1.0, and some library still depends on 0.2, and `custom-impl` is not set, the library would quietly fallback to the default impl, which might not be multi-core safe.

(At the same time it's very desirable that a HAL may implement 0.2 and 1.0 at the same time, to allow for smooth migrations. See https://github.com/rp-rs/rp-hal/pull/404 for an implementation supporting both 0.2 and 1.0)

**Comment by jannic** [comment]
In fact, I'd very much prefer a compile error due to duplicate symbols (which could be properly explained in the docs) compared to a silent fallback to the 0.2 default implementation.

**Comment by jannic** [comment]
How about https://github.com/embassy-rs/critical-section/pull/16 ?

**Comment by jannic** [comment]
To summarize #16:
- if the 1.0 `RestoreState` fits inside the `u8` restore state of 0.2, also provide the 0.2 low level primitives with their old names, `_critical_section_acquire`, `_critical_section_release`. So other crates still linking to critical-section 0.2 just keep working, correctly.
- if the 1.0 `RestoreState` is larger than an `u8`, this is not possible. I guess in most of these cases, there won't be a critical-section 0.2 implementation for that target at all, because why choose a larger state if a smaller one works as well? Still, to be safe, define `_critical_section_acquire` and `_critical_section_release`, but make them in turn depend on a non-existing symbol `_do_not_use_critical_section_version_0_2`. If these functions are never called, this doesn't matter. The linker will just remove those functions. But if some other crate does use critical-section 0.2, those functions will be referenced, and a linker error will occur, preventing a subtly broken binary.

In essence, the most common case just works, and silently non-working critical sections are prevented. Non-working combinations cause compile-time errors.

The only disadvantage I currently see is that the second case depends on the linker properly removing the unused symbols, before complaining about the undefined ones. It seems to work, I tested a few combinations. But may this break with other linkers, or even different linker settings? If this doesn't work reliably, it could be an alternative to just panic inside those functions. This will make it an (hopefully obvious) runtime error instead of a compile time error, though.

**Comment by Dirbaio** [comment]
it's an interesting solution indeed :thinking: 

> So other crates still linking to critical-section 0.2 just keep working, correctly.

If 0.2's `custom-impl` feature is not enabled, the 0.2 crate will try to supply the default impl, which will result in linker errors. The only way to fix it is to make 1.0 depend on 0.2 to enable `custom-impl`, and I really don't like that. I'd rather have 0.2 go completely unused forever, perhaps even yank it in the future when everyone's moved to 1.0.

> the second case depends on the linker properly removing the unused symbols

I've seen this trick used before, for example in `panic-never`. I think relying on the linker removing unused stuff is reasonable in practice. We can try it, and if someone reports problems we can always change it to panicking, because it wouldn't be a breaking change.

An alternative would be to always "sabotage" 0.2? ie, always doing what your PR does in the non-u8 case, so that you always get a linker error if you mix 0.2 + 1.0.

**Comment by jannic** [comment]
I'd prefer if a HAL (or whatever provides the implementation) could support both 1.0 and 0.2 at the same time, to avoid requiring all crates to be updated at once.
So actively sabotaging 0.2 would be unfortunate.
Such a HAL would enable 0.2's custom-impl feature without actually providing an implementation, relying on the implementation provided by 1.0.

**Comment by Dirbaio** [comment]
Look at the [dependents of CS0.2 on crates.io](https://crates.io/crates/critical-section/reverse_dependencies). There's just 5 widely-used crates.

- atomic-polyfill
- defmt-rtt
- esp-idf-hal
- rp2040-hal
- nrf-usbd

People in this thread have release access to 4 of these, and I'm sure the defmt folks will get a defmt-rtt 0.4 release with it quite fast. We can get everyone updated really quick!

I honestly think it's not worth carrying legacy hacks in CS1.0 forever.

**Comment by jannic** [comment]
I agree that updating these crates will happen quickly. But some of these are in turn dependencies of a lot of other crates. So an old version of those may still cause critical-section 0.2 to be loaded.
And updating defmt is an topic of it's own. I had issues with that a few days ago. https://github.com/rp-rs/rp-hal/pull/398, still not solved.

**Comment by Dirbaio** [comment]
`defmt 0.2 -> defmt 0.3` is a pain because you can't mix them in the same binary. This would need bumping just `defmt-rtt 0.4`, it  would keep using `defmt 0.3`. Only the end user binary depends on `defmt-rtt`, libs only depend on `defmt`.

**Comment by Dirbaio** [comment]
0.2 compat was discussed a bit more and decided against on this week's [WG meeting](https://github.com/rust-embedded/wg/blob/master/minutes/2022-08-09.md).

**Comment by Dirbaio** [comment]
`critical-section` 1.0.0 is now released! https://docs.rs/critical-section/1.0.0/critical_section/index.html
